python/ray/air/preprocessor.py

```diff
@@ -165,18 +166,62 @@ def _fit(self, dataset: Dataset) -> "Preprocessor":
-        return dataset.map_batches(self._transform_pandas, batch_format="pandas")
+        dataset_format = dataset._dataset_format()
+        exc_msg = "Neither `_transform_arrow` nor `_transform_pandas` are implemented."
+
+        # If only _transform_arrow is implemented, will convert the data to arrow.
+        # If only _transform_pandas is implemented, will convert the data to pandas.
+        # If both are implemented, will pick the method corresponding to the format
+        # for best performance.
+        # Implementation is defined as overriding the method in a sub-class.
+        if dataset_format == "arrow":
+            if self.__class__._transform_arrow != Preprocessor._transform_arrow:
+                return dataset.map_batches(
+                    self._transform_arrow, batch_format="pyarrow"
+                )
+            elif self.__class__._transform_pandas != Preprocessor._transform_pandas:
+                return dataset.map_batches(
+                    self._transform_pandas, batch_format="pandas"
+                )
+            else:
+                raise NotImplementedError(exc_msg)
+        elif dataset_format == "pandas":
+            if self.__class__._transform_pandas != Preprocessor._transform_pandas:
+                return dataset.map_batches(
+                    self._transform_pandas, batch_format="pandas"
+                )
+            elif self.__class__._transform_arrow != Preprocessor._transform_arrow:
+                return dataset.map_batches(
+                    self._transform_arrow, batch_format="pyarrow"
+                )
+            else:
+                raise NotImplementedError(exc_msg)
+        else:
+            raise ValueError(
+                f"Unsupported Dataset format: '{dataset_format}'. Only 'pandas' and "
+                "'arrow' Dataset formats are supported."
+            )
 
     def _transform_batch(self, df: DataBatchType) -> DataBatchType:
         import pandas as pd
 
-        # TODO(matt): Add `_transform_arrow` to use based on input type.
-        # Reduce conversion cost if input is in Arrow:  Arrow -> Pandas -> Arrow.
-        if not isinstance(df, pd.DataFrame):
+        try:
+            import pyarrow
+        except ImportError:
+            pyarrow = None
+
+        if isinstance(df, pd.DataFrame):
+            return self._transform_pandas(df)
+        elif pyarrow is not None and isinstance(df, pyarrow.Table):
+            return self._transform_arrow(df)
+        else:
             raise NotImplementedError(
-                "`transform_batch` is currently only implemented for Pandas DataFrames."
+                "`transform_batch` is currently only implemented for Pandas DataFrames "
+                f"and PyArrow Tables. Got {type(df)}."
             )
-        return self._transform_pandas(df)
 
     def _transform_pandas(self, df: "pd.DataFrame") -> "pd.DataFrame":
         raise NotImplementedError()
+
```

Comment:
Can you mark these as DeveloperAPI?

Reply:
I thought we don't mark individual methods? They are already underscored, too

Reply:
No we do mark individual methods- see Trainer for example.

The conventions for underscores are a bit weird, but I would like to explicitly mark the methods that people implementing new preprocessors need to override as Developer APIs

Reply:
Also can we add docstrings for these?